sources/rasterio/large_image_source_rasterio/__init__.py

```diff
@@ -0,0 +1,1078 @@
+try:
+    __version__ = _importlib_version(__name__)
+except PackageNotFoundError:
+    # package is not installed
+    pass
+
+TileInputUnits['projection'] = 'projection'
+TileInputUnits['proj'] = 'projection'
+TileInputUnits['wgs84'] = 'proj4:EPSG:4326'
+TileInputUnits['4326'] = 'proj4:EPSG:4326'
+
+# Inform the tile source cache about the potential size of this tile source
+CacheProperties['tilesource']['itemExpectedSize'] = max(
+    CacheProperties['tilesource']['itemExpectedSize'], 100 * 1024**2
+)
+
+# Used to cache pixel size for projections
+ProjUnitsAcrossLevel0 = {}
+ProjUnitsAcrossLevel0_MaxSize = 100
+
+
+class RasterioFileTileSource(GDALBaseFileTileSource, metaclass=LruCacheMetaclass):
+    """Provides tile access to geospatial files."""
+
+    cacheName = 'tilesource'
+    name = 'rasterio'
+    extensions = {
+        None: SourcePriority.MEDIUM,
+        'geotiff': SourcePriority.PREFERRED,
+        'ntf': SourcePriority.PREFERRED,
+        'nitf': SourcePriority.PREFERRED,
+        'tif': SourcePriority.LOW,
+        'tiff': SourcePriority.LOW,
+        'vrt': SourcePriority.PREFERRED,
+    }
+    mimeTypes = {
+        None: SourcePriority.FALLBACK,
+        'image/geotiff': SourcePriority.PREFERRED,
+        'image/tiff': SourcePriority.LOW,
+        'image/x-tiff': SourcePriority.LOW,
+    }
+
+    def __init__(self, path, projection=None, unitsPerPixel=None, **kwargs):
+        """Initialize the tile class.
+
+        See the base class for other available parameters.
+
+        :param path: a filesystem path for the tile source.
+        :param projection: None to use pixel space, otherwise a crs compatible with pyproj.
+        :param unitsPerPixel: The size of a pixel at the 0 tile size.
+            Ignored if the projection is None.  For projections, None uses the default,
+            which is the distance between (-180,0) and (180,0) in EPSG:4326 converted to the
+            projection divided by the tile size. crs projections that are not latlong
+            (is_geographic is False) must specify unitsPerPixel.
+
+        """
+        # init the object
+        super().__init__(path, **kwargs)
+
+        # create a thred lock
+        self._getDatasetLock = threading.RLock()
+
+        if isinstance(path, rio.io.MemoryFile):
+            path = path.open(mode='r')
+
+        if isinstance(path, rio.io.DatasetReaderBase):
+            self.dataset = path
+            self._largeImagePath = self.dataset.name
+        else:
+            # set the large_image path
+            self._largeImagePath = self._getLargeImagePath()
+
+            # open the file with rasterio and display potential warning/errors
+            with self._getDatasetLock:
+                try:
+                    self.dataset = rio.open(self._largeImagePath)
+                except RasterioIOError:
+                    raise TileSourceError('File cannot be opened via rasterio.')
+                except FileNotFoundError:
```

Comment:
When the file doesn't exist, this isn't the code path it is following.  Maybe in the section where we catch RasterIOError, we add
```
            if not os.path.isfile(self._largeImagePath):
                raise TileSourceFileNotFoundError(self._largeImagePath) from None
```